From f54c99b2ae12f7738f0c5265c498a6ef51bf8399 Mon Sep 17 00:00:00 2001
From: Fujii Hironori <fujii@igalia.com>
Date: Mon, 22 Sep 2025 04:22:54 -0700
Subject: Cherry-pick 300316@main (2b754f0fec28).
 https://bugs.webkit.org/show_bug.cgi?id=298599

    wpe_view_render_buffer: assertion 'WPE_IS_BUFFER(buffer)' failed
    https://bugs.webkit.org/show_bug.cgi?id=298599

    Reviewed by Carlos Garcia Campos.

    An assertion failed in wpe_view_render_buffer() if the active web
    process was switched before the previous rendering request wasn't
    finished.

    > ** (MiniBrowser:190744): CRITICAL **: 23:35:59.702: gboolean wpe_view_render_buffer(WPEView*, WPEBuffer*, const WPERectangle*, guint, GError**): assertion 'WPE_IS_BUFFER(buffer)' failed

    In AcceleratedBackingStore class, if updateSurfaceID() is called
    before renderPendingBuffer() is called, m_pendingBuffer is null in
    renderPendingBuffer(). Check if m_pendingBuffer is null in the fence
    monitor callback.

    * Source/WebKit/UIProcess/wpe/AcceleratedBackingStore.cpp:
    (WebKit::AcceleratedBackingStore::AcceleratedBackingStore):

    Canonical link: https://commits.webkit.org/300316@main

Canonical link: https://commits.webkit.org/298234.140@webkitglib/2.50
---
 Source/WebKit/UIProcess/wpe/AcceleratedBackingStore.cpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

Upstream-Status: Backport [https://bugs.webkit.org/show_bug.cgi?id=298599]

diff --git a/Source/WebKit/UIProcess/wpe/AcceleratedBackingStore.cpp b/Source/WebKit/UIProcess/wpe/AcceleratedBackingStore.cpp
index 8aa136cef4b4..c4a0d262b967 100644
--- a/Source/WebKit/UIProcess/wpe/AcceleratedBackingStore.cpp
+++ b/Source/WebKit/UIProcess/wpe/AcceleratedBackingStore.cpp
@@ -52,7 +52,10 @@ Ref<AcceleratedBackingStore> AcceleratedBackingStore::create(WebPageProxy& webPa
 AcceleratedBackingStore::AcceleratedBackingStore(WebPageProxy& webPage, WPEView* view)
     : m_webPage(webPage)
     , m_wpeView(view)
-    , m_fenceMonitor([this] { renderPendingBuffer(); })
+    , m_fenceMonitor([this] {
+        if (m_webPage && m_pendingBuffer)
+            renderPendingBuffer();
+    })
     , m_legacyMainFrameProcess(webPage.legacyMainFrameProcess())
 {
     g_signal_connect(m_wpeView.get(), "buffer-rendered", G_CALLBACK(+[](WPEView*, WPEBuffer*, gpointer userData) {
-- 
2.43.0

